src/components/App.tsx

```diff
@@ -69,7 +92,21 @@ export default class App extends PureComponent<Props> {
 								write: this.props.writeToStderr
 							}}
 						>
-							{this.props.children}
+							<FocusContext.Provider
+								value={{
+									activeId: this.state.activeFocusId,
+									add: this.addFocusable,
+									remove: this.removeFocusable,
+									activate: this.activateFocusable,
+									deactivate: this.deactivateFocusable,
+									enableFocus: this.enableFocus,
+									disableFocus: this.disableFocus,
+									focusNext: this.focusNext,
+									focusPrevious: this.focusPrevious
+								}}
+							>
```

Comment:
Personally, I'd move this context provider and all the following methods into a separate file because this file becomes more complicated with those 200 supplementary lines. (ie. someone that want to improve ink might be lost in those lines whereas he does not need to understand focus management)

Reply:
Agree and this is what I wanted to do initially, but decided to leave it in `App.tsx` for now to reuse the `data` event handler on `stdin` stream, instead of having two separate ones. I think a better move here is to rewrite this file using hooks and extract individual functionality (focus management, exiting, raw mode) into hooks, which will clean this whole thing up.